### ISSUE TITLE
Adjust dropout rate over epochs

### ISSUE THREAD

**Issue by bobchennan** [fchollet/keras#3424]
[Some research papers](http://www.clsp.jhu.edu/~samuel/pdfs/annealed_dropout.pdf) showed the benefit to adjust dropout rate over epochs.
Specifically, they used this dropout rate:

    max(0,cur/N)*initial_dropoutrate

where ```cur``` is the current epoch and N is a fixed parameter.

How can I achieve this using Keras? I think I can re-write the call function in the dropout layer but the problem is that how can I get the epoch variable.

**Comment by ncullen93** [comment]
make your own layer that changes the dropout probability based on whatever condition you want.. Start from the Dropout class itself is probably a good idea.. you can see how 'self.p' can easily be altered in the class or in a separate callback.

class Dropout(Layer):
    '''Applies Dropout to the input. Dropout consists in randomly setting
    a fraction `p` of input units to 0 at each update during training time,
    which helps prevent overfitting.

    # Arguments
        p: float between 0 and 1. Fraction of the input units to drop.

    # References
        - [Dropout: A Simple Way to Prevent Neural Networks from Overfitting](http://www.cs.toronto.edu/~rsalakhu/papers/srivastava14a.pdf)
    '''
    def __init__(self, p, **kwargs):
        self.p = p
        if 0. < self.p < 1.:
            self.uses_learning_phase = True
        self.supports_masking = True
        super(Dropout, self).__init__(**kwargs)

    def call(self, x, mask=None):
        if 0. < self.p < 1.:
            x = K.in_train_phase(K.dropout(x, level=self.p), x)
        return x

    def get_config(self):
        config = {'p': self.p}
        base_config = super(Dropout, self).get_config()
        return dict(list(base_config.items()) + list(config.items()))

**Comment by bobchennan** [comment]
Yes, I implemented it using callback and a new shared variable.
Also a little hack is needed since function $K.dropout$ check whether the level is in range 0-1 or not. 
In this case the level is a function instead of a constant.

**Comment by oak-tree** [comment]
Hey @ncullen93 
Quick question: How can I change the `parameter rate` of the `Dropout` class from a `callback` function? It cannot find any reference the callback can use to reach it.

Thanks!

**Comment by bobchennan** [comment]
In the callback function the model is given and you can iterator over the
layers.

**Comment by oak-tree** [comment]
I'll checkout it, thanks!

**Comment by edwardyu** [comment]
Overwriting the `call` function is not working for me because the `call` function is not called on every forward pass. Any ideas? 
```
def call(self, inputs, training=None):
        print("In call function.")
        def noised():
            print("noised()")
            stddev = self.t * self.stddev
            return inputs * K.random_normal(shape=K.shape(inputs),
                                            mean=1.0,
                                            stddev=stddev)

        return K.in_train_phase(noised, inputs, training=training)
```
I increment `self.t` in a callback function.

**Comment by kilean20** [comment]
Hi, but changing the dropout rate (here Dropout.p) through keras.callback.Callback did not work for me. I think the changed rate is not reflected in the compiled model.
